src/realm/query_conditions.hpp

```diff
@@ -65,12 +62,65 @@ class QueryStateBase {
     {
     }
 
+    // Called when we have a match.
+    // The return value indicates if the query should continue.
+    virtual bool match(size_t, Mixed) noexcept = 0;
+
+    virtual bool match_pattern(size_t, uint64_t)
+    {
+        return false;
+    }
+
+protected:
 private:
     virtual void dyncast();
 };
 
 template <class>
-class QueryState;
+class QueryStateMin;
+
+template <class>
+class QueryStateMax;
+
+class QueryStateCount : public QueryStateBase {
+public:
+    QueryStateCount(size_t limit = -1)
+        : QueryStateBase(limit)
+    {
+    }
+    bool match(size_t, Mixed) noexcept final
+    {
+        ++m_match_count;
+        return (m_limit > m_match_count);
+    }
+    size_t get_count() const noexcept
+    {
+        return m_match_count;
+    }
+};
+
+namespace aggregate_operations {
+
+template <typename T>
+static bool is_nan(T value)
+{
+    if constexpr (std::is_floating_point_v<T>) {
+        return std::isnan(value);
+    }
+    else {
+        // gcc considers the argument unused if it's only used in one branch of if constexpr
+        static_cast<void>(value);
+        return false;
+    }
+}
+
+template <>
+inline bool is_nan<Decimal128>(Decimal128 value)
+{
+    return value.is_nan();
+}
```

Comment:
Whoops, my comment got lost somewhere: Why the mix of specialization and `if constexpr`?